crates/samples/components/json_validator/src/lib.rs

```diff
@@ -0,0 +1,189 @@
+use jsonschema::JSONSchema;
+use windows::{core::*, Win32::Foundation::*};
+
+// Creates a JSON validator object with the given schema. The returned handle must be freed
+// by calling `CloseJsonValidator`.
+#[no_mangle]
+unsafe extern "system" fn CreateJsonValidator(
+    schema: *const u8,
+    schema_len: usize,
+    handle: *mut usize,
+) -> HRESULT {
+    create_validator(schema, schema_len, handle).into()
+}
+
+// Validates a JSON value against a previously-compiled schema.
+#[no_mangle]
+unsafe extern "system" fn ValidateJson(
+    handle: usize,
+    value: *const u8,
+    value_len: usize,
+) -> HRESULT {
+    validate(handle, value, value_len).into()
+}
+
+// Closes a JSON validator object.
+#[no_mangle]
+unsafe extern "system" fn CloseJsonValidator(handle: usize) {
+    if handle != 0 {
+        _ = Box::from_raw(handle as *mut JSONSchema);
+    }
+}
+
+// Implementation of the `CreateJsonValidator` function so we can use `Result` for simplicity.
+unsafe fn create_validator(schema: *const u8, schema_len: usize, handle: *mut usize) -> Result<()> {
+    let schema = json_from_raw_parts(schema, schema_len)?;
+
+    let compiled = JSONSchema::compile(&schema)
+        .map_err(|error| Error::new(E_INVALIDARG, error.to_string().into()))?;
+
+    if handle.is_null() {
+        return Err(E_POINTER.into());
+    }
+
+    // The handle is not null so we can safely dereference it here.
+    *handle = Box::into_raw(Box::new(compiled)) as usize;
+
+    Ok(())
+}
+
+// Implementation of the `ValidateJson` function so we can use `Result` for simplicity.
+unsafe fn validate(handle: usize, value: *const u8, value_len: usize) -> Result<()> {
```

Comment:
Our current helper has a "return a reconstructed json from the parsed DOM" model so that the downstream consumer is only touching "produced by code we own, safely" bits. Is there an example of returning a string here?

Reply:
Sure, `HSTRING` would be simple but it probably makes more sense to return a UTF-8 string that would need to be freed with something like `HeapFree` or `CoTaskMemFree`.

Reply:
I've updated the PR with such an example. https://github.com/microsoft/windows-rs/pull/2815/commits/864de7961db415b065d3c36c6bb02cdc8d85dc1c